rust/x509/src/x509_example_rust/src/wasm_only/signer.rs

```diff
@@ -0,0 +1,142 @@
+                (internal_request,),
+                26_153_846_153,
+            )
+            .await
+            .map_err(|e| format!("sign_with_schnorr failed {e:?}"))?;
+
+        Ok(internal_reply.signature)
+    }
+}
+
+impl AsRef<ed25519::pkcs8::PublicKeyBytes> for Ed25519Signer {
+    fn as_ref(&self) -> &ed25519::pkcs8::PublicKeyBytes {
+        &self.public_key
+    }
+}
+
+impl signature::KeypairRef for Ed25519Signer {
+    type VerifyingKey = ed25519::pkcs8::PublicKeyBytes;
+}
+
+impl DynSignatureAlgorithmIdentifier for Ed25519Signer {
+    fn signature_algorithm_identifier(&self) -> spki::Result<AlgorithmIdentifierOwned> {
+        Ok(AlgorithmIdentifierOwned {
+            oid: ed25519::pkcs8::ALGORITHM_OID.clone(),
+            parameters: None,
+        })
+    }
+}
+
+pub struct EcdsaSecp256k1Signer {
+    key_id: cdk_ecdsa::EcdsaKeyId,
+    public_key: k256::ecdsa::VerifyingKey,
+}
+
+impl EcdsaSecp256k1Signer {
+    pub async fn new() -> Result<Self, String> {
+        let public_key = k256::ecdsa::VerifyingKey::from_encoded_point(
+            &k256::PublicKey::from_sec1_bytes(&root_ca_public_key_bytes().await?.as_slice())
+                .map_err(|e| format!("malformed public key: {e:?}"))?
+                .to_encoded_point(false),
+        )
+        .unwrap();
+        Ok(Self {
+            key_id: CA_KEY_INFORMATION
+                .with(|value| cdk_ecdsa::EcdsaKeyId::try_from(value.borrow().deref()))?,
+            public_key,
+        })
+    }
+}
+
+impl Sign for EcdsaSecp256k1Signer {
+    async fn sign(&self, msg: &[u8]) -> Result<Vec<u8>, String> {
+        let mut hasher = sha2::Sha256::new();
+        hasher.update(msg);
+
+        let args = cdk_ecdsa::SignWithEcdsaArgument {
+            message_hash: hasher.finalize().to_vec(),
+            derivation_path: derivation_path(),
+            key_id: self.key_id.clone(),
+        };
+
+        let (internal_reply,): (cdk_ecdsa::SignWithEcdsaResponse,) =
+            cdk_ecdsa::sign_with_ecdsa(args)
+                .await
+                .map_err(|e| format!("sign_with_schnorr failed {e:?}"))?;
+
+        Ok(
+            k256::ecdsa::Signature::from_slice(internal_reply.signature.as_slice())
+                .map_err(|e| format!("malformed signature: {e:?}"))?
+                .to_der()
+                .to_bytes()
+                .to_vec(),
+        )
+    }
+}
+
+impl AsRef<k256::ecdsa::VerifyingKey> for EcdsaSecp256k1Signer {
+    fn as_ref(&self) -> &k256::ecdsa::VerifyingKey {
+        &self.public_key
+    }
+}
+
+impl signature::KeypairRef for EcdsaSecp256k1Signer {
+    type VerifyingKey = k256::ecdsa::VerifyingKey;
+}
+
+impl DynSignatureAlgorithmIdentifier for EcdsaSecp256k1Signer {
+    fn signature_algorithm_identifier(&self) -> spki::Result<AlgorithmIdentifierOwned> {
+        Ok(AlgorithmIdentifierOwned {
+            oid: ecdsa::ECDSA_SHA256_OID,
+            parameters: Some(der::Any::from(k256::Secp256k1::OID)),
```

Comment:
My read of RFC 5758 is that the parameters should not be included here. (May well be the case that most implementations just ignore the parameters however)

Reply:
It is IMO okay to not use those parameters at all and reject certificates using them. This is either way not a full implementation but only a demo.